eclipse-glsp/glsp-client
# Improve handling of feedback emitters and fix some minor bugs

#### What it does
- Provide dedicated feedback emitter object as stable emitter
- Allow cleanup functions to be calculated as a lazy function
- Log if a potentially unstable feedback emitter is used
- Split feedback in change bounds tool into dedicated files
- Properly use feedback emitter to cleanup feedback correctly

Fixes https://github.com/eclipse-glsp/glsp/issues/1323 Fixes https://github.com/eclipse-glsp/glsp/issues/1243

- Ensure correct snapping of ghost element position on grid

Fixes https://github.com/eclipse-glsp/glsp/issues/1256



#### How to test

- Check the current master and this PR to compare feedback handling.
- For particular issues, check the issue description.

#### Follow-ups

#### Changelog


Feedback is strongly tied to an emitter that serves as a key for the particular actions that make up the feedback. While any object can serve as and emitter, feedback de-registration and cleanup depend on it.  As such it is very important that the emitter is a stable object and not, for instance, a model element which changes with each update. While previously, we often used the tool or mouse listener themselves as emitters, this change introduces a dedicated feedback emitter object that encapsulates a set of feedback and cleanup actions and provides convenience methods to build up that feedback gradually. This is especially useful for providing hooks to add additional feedback or for tools that provide several feedbacks that are not cleaned and disposed together. Our existing tools and feedback listeners have been updated to use those emitter objects as a best practice.
